{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pandas\n",
    "\n",
    "## What is Pandas?\n",
    "\n",
    "**Python Data Analysis Library**\n",
    "\n",
    "A powerful library for manipulating data arranged in formats like matricies and data frames. It's arguably the most popular Python library used for data analysis. Pandas aims to provide Python users the same type of functionality as the popular statistical language, **R**.\n",
    "\n",
    "<br/>\n",
    "\n",
    "![](https://d2h0cx97tjks2p.cloudfront.net/blogs/wp-content/uploads/sites/2/2019/04/Python-Pandas-Applications.jpg)\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Why familiarize yourself with Pandas?\n",
    "\n",
    "So far we've discussed building multidimensional objects like lists of lists and dictionaries of dictionaries from raw data. However, bioinformatics modules (and many others) will often **return** results in the form of Pandas **data frame** or **a matrix**. Further manipulation of these results will require some degree of Pandas operations.\n",
    "\n",
    "For example, lets say you want to parse your RNA-seq results to a list of genes within a specific range of p-values and log fold changes, like \"all p-values less than `1e-15` and log fold changes greater than `1.2`\". You can apply your knowledge of Python operators such as `and, >, <` to subset a data frame based on the afformentioned parameters.\n",
    "\n",
    "<br/>\n",
    "\n",
    "## Types of data manipulated in Pandas\n",
    "\n",
    "### Matrices\n",
    "\n",
    "A matrix is an data structure where numbers are arranged into rows and columns. They will typically conatin floats __or__ integers, but not both. Matrices are used when you need to perform mathmatical operations between datasets that contain multiple dimensions.\n",
    "\n",
    "<br/>\n",
    "\n",
    "![](https://upload.wikimedia.org/wikipedia/commons/thumb/2/26/Gene_co-expression_network_construction_steps.png/720px-Gene_co-expression_network_construction_steps.png)\n",
    "\n",
    "<br/>\n",
    "\n",
    "### Data frames\n",
    "\n",
    "A data frame is a table-like data structure and can cotain mixed data types (strings, floats, integers, etc.). This is the type of data structure you're used seeing in Excel.\n",
    "\n",
    "<br/>\n",
    "\n",
    "![](https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0161567.t005&type=large)\n",
    "\n",
    "<br/>\n",
    "\n",
    "## Pandas has the ability to read in various data formats\n",
    "\n",
    "- Open a local file using Pandas, usually a CSV file, but could also be a delimited text file (like TSV), Excel, etc\n",
    "\n",
    "- Open a remote file or database like a CSV or a JSONon a website through a URL or read from a SQL table/database\n",
    "\n",
    "<br/>\n",
    "\n",
    "## A brief word on vectorization\n",
    "\n",
    "**Operations in Pandas, like R, work most efficiently when vectorized**<br/>\n",
    "\n",
    "<br/>\n",
    "\n",
    "You can think of a vector (also reffered to as an [array](https://docs.python.org/3/library/array.html)) as a list that contains a single data type used. For matricies, this can be any row or column.\n",
    "\n",
    "Rather that looping through individual values (scalars), we _apply_ operations to vectors. That is, the vector is treated as a single object. \n",
    "\n",
    "If you frequently work dataframes or matricies, please consider reading this \n",
    "[article](https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6) and this\n",
    "[stackoverflow thread](https://stackoverflow.com/questions/35091979/why-is-vectorization-faster-in-general-than-loops) for a detailed explanation.\n",
    "\n",
    "<br/>\n",
    "\n",
    "<img src=\"https://miro.medium.com/max/2060/1*p4zjrqG97C4bFmOXU5UQog.png\" width=\"80%\" height=\"80%\" />\n",
    "\n",
    "<br/>\n",
    "\n",
    "## Basic methods for data manipulation\n",
    "\n",
    "### Reading in csv files and row/column slicing\n",
    "\n",
    "\"Slicing\" refers to subsetting, or removing rows and columns from a data frame. Here we'll read in a data frame, look at the contents, and subset it by slicing out arbitrary regions.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Setting index_col to 0 tells us that the first column contains the row names\n",
    "cell_attributes = pd.read_csv(\"./meta_data.csv\", index_col = 0)\n",
    "\n",
    "type(cell_attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       n_counts  n_genes  orig_ident  percent_mito  res_2  \\\n",
      "index                                                                       \n",
      "AAACCTGAGCACCGCT-1_1       1980      928           1      0.026781      1   \n",
      "AAACCTGCACATTCGA-1_2       2238      894           1      0.024150      6   \n",
      "AAACCTGCAGGGTTAG-1_3       2839     1025           1      0.005640      7   \n",
      "AAACCTGCATGCCTTC-1_4       3437     1438           1      0.012529     10   \n",
      "AAACCTGGTTGAACTC-1_5       2765     1277           1      0.010138      9   \n",
      "AAACCTGTCACCTTAT-1_6       2771     1144           1      0.019530      1   \n",
      "AAACGGGTCCACGTGG-1_8       5586     2022           1      0.006445      4   \n",
      "AAAGATGCAAGGACTG-1_9       1904      831           1      0.025223      2   \n",
      "AAAGATGCACGAAATA-1_10      1879      822           1      0.019702      4   \n",
      "AAAGATGCATCGATTG-1_11      1135      566           1      0.056537      1   \n",
      "AAAGATGGTACACCGC-1_12      1181      542           1      0.034716      3   \n",
      "AAAGATGGTCTTTCAT-1_13      2540      982           1      0.027964      3   \n",
      "AAAGCAAAGATGCCAG-1_14      3953     1414           1      0.013411      3   \n",
      "AAAGCAACACAGATTC-1_15      3206     1263           1      0.009054      8   \n",
      "AAAGCAACACGGCGTT-1_16      1355      563           1      0.056911     10   \n",
      "AAAGCAATCGTAGATC-1_18      1647      830           1      0.019465      0   \n",
      "AAAGTAGAGAATAGGG-1_19      1187      544           1      0.021922      2   \n",
      "AAAGTAGAGGTCATCT-1_20      4106     1527           1      0.017317     11   \n",
      "AAAGTAGGTAAGTGGC-1_21      1217      545           1      0.047658      0   \n",
      "AAAGTAGGTGCAGACA-1_22      3671     1326           1      0.045244      5   \n",
      "AAAGTAGTCACCACCT-1_23      6873     1876           1      0.018055      5   \n",
      "AAATGCCCACTGCCAG-1_24      1478      680           1      0.078751      5   \n",
      "AAATGCCGTATTCGTG-1_25      3129     1107           1      0.017589     13   \n",
      "AACACGTCAAGCGATG-1_26      1460      671           1      0.030864      7   \n",
      "AACACGTGTATCAGTC-1_27      2586     1150           1      0.006961      2   \n",
      "\n",
      "                       tree_ident  louvain  dpt_pseudotime  \n",
      "index                                                       \n",
      "AAACCTGAGCACCGCT-1_1            7        0        0.353522  \n",
      "AAACCTGCACATTCGA-1_2            9        2        0.177180  \n",
      "AAACCTGCAGGGTTAG-1_3           14        8        0.000000  \n",
      "AAACCTGCATGCCTTC-1_4           12        1        0.103049  \n",
      "AAACCTGGTTGAACTC-1_5            6       10        0.259682  \n",
      "AAACCTGTCACCTTAT-1_6            7        0        0.316132  \n",
      "AAACGGGTCCACGTGG-1_8            3        9        0.341969  \n",
      "AAAGATGCAAGGACTG-1_9           13        1        0.114736  \n",
      "AAAGATGCACGAAATA-1_10           3        9        0.431574  \n",
      "AAAGATGCATCGATTG-1_11           7        0        0.247803  \n",
      "AAAGATGGTACACCGC-1_12           8        2        0.147456  \n",
      "AAAGATGGTCTTTCAT-1_13           8        6        0.073615  \n",
      "AAAGCAAAGATGCCAG-1_14           8        6        0.095377  \n",
      "AAAGCAACACAGATTC-1_15           5        7        0.558053  \n",
      "AAAGCAACACGGCGTT-1_16          12       13        0.126446  \n",
      "AAAGCAATCGTAGATC-1_18          11        4        0.046991  \n",
      "AAAGTAGAGAATAGGG-1_19          13        1        0.125822  \n",
      "AAAGTAGAGGTCATCT-1_20           2       11        0.438730  \n",
      "AAAGTAGGTAAGTGGC-1_21          11        4        0.062753  \n",
      "AAAGTAGGTGCAGACA-1_22           1        5        0.979432  \n",
      "AAAGTAGTCACCACCT-1_23           1        5        0.883077  \n",
      "AAATGCCCACTGCCAG-1_24           1        5        0.823691  \n",
      "AAATGCCGTATTCGTG-1_25           4        8        0.067783  \n",
      "AACACGTCAAGCGATG-1_26          14        8        0.021948  \n",
      "AACACGTGTATCAGTC-1_27          13        1        0.126628  \n"
     ]
    }
   ],
   "source": [
    "# We notice rows and columns are truncated with the dimensions given the bottom\n",
    "print(cell_attributes.head(25))\n",
    "\n",
    "# Change the output view options\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_columns', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<br/>\n",
    "\n",
    "#### loc/iloc\n",
    "\n",
    "Pandas has different methods for subsetting dataframes.\n",
    "We'll dicuss the most common methods, **loc**, and **iloc**\n",
    "\n",
    "loc allows us to subset data by row or column label. For example, if I would\n",
    "like to subset the column 'n_counts', I would use the following command:\n",
    "\n",
    "<br/>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "AAACCTGAGCACCGCT-1_1       1980\n",
       "AAACCTGCACATTCGA-1_2       2238\n",
       "AAACCTGCAGGGTTAG-1_3       2839\n",
       "AAACCTGCATGCCTTC-1_4       3437\n",
       "AAACCTGGTTGAACTC-1_5       2765\n",
       "AAACCTGTCACCTTAT-1_6       2771\n",
       "AAACGGGTCCACGTGG-1_8       5586\n",
       "AAAGATGCAAGGACTG-1_9       1904\n",
       "AAAGATGCACGAAATA-1_10      1879\n",
       "AAAGATGCATCGATTG-1_11      1135\n",
       "AAAGATGGTACACCGC-1_12      1181\n",
       "AAAGATGGTCTTTCAT-1_13      2540\n",
       "AAAGCAAAGATGCCAG-1_14      3953\n",
       "AAAGCAACACAGATTC-1_15      3206\n",
       "AAAGCAACACGGCGTT-1_16      1355\n",
       "AAAGCAATCGTAGATC-1_18      1647\n",
       "AAAGTAGAGAATAGGG-1_19      1187\n",
       "AAAGTAGAGGTCATCT-1_20      4106\n",
       "AAAGTAGGTAAGTGGC-1_21      1217\n",
       "AAAGTAGGTGCAGACA-1_22      3671\n",
       "AAAGTAGTCACCACCT-1_23      6873\n",
       "AAATGCCCACTGCCAG-1_24      1478\n",
       "AAATGCCGTATTCGTG-1_25      3129\n",
       "AACACGTCAAGCGATG-1_26      1460\n",
       "AACACGTGTATCAGTC-1_27      2586\n",
       "AACCATGTCACCCTCA-1_28      3823\n",
       "AACCGCGAGTGTTGAA-1_29      1432\n",
       "AACGTTGAGGTGCTTT-1_30      9787\n",
       "AACGTTGCACGGTAAG-1_31      4425\n",
       "AACGTTGGTTCGGCAC-1_32      4266\n",
       "AACTCAGCACAGACAG-1_33      8168\n",
       "AACTCAGGTAAGGGCT-1_34      1682\n",
       "AACTCCCAGCGCTCCA-1_36      3447\n",
       "AACTCCCTCGCGATCG-1_38      6047\n",
       "AACTCTTAGTAGTGCG-1_39      3261\n",
       "AACTCTTGTCTTCGTC-1_40      2720\n",
       "AACTGGTCAGCCAGAA-1_42      4785\n",
       "AACTGGTCAGGTCGTC-1_43      6532\n",
       "AACTTTCGTTTACTCT-1_45      1160\n",
       "AACTTTCTCAGAGCTT-1_46      3489\n",
       "AACTTTCTCGTACCGG-1_47      1513\n",
       "AAGACCTAGAATAGGG-1_48      8797\n",
       "AAGACCTAGGAGCGAG-1_49      4993\n",
       "AAGACCTCAGAGCCAA-1_50      6268\n",
       "AAGCCGCAGCTTTGGT-1_51      5254\n",
       "AAGCCGCAGGAGCGAG-1_52      5309\n",
       "AAGCCGCCACATGGGA-1_53      1364\n",
       "AAGCCGCGTTCGCTAA-1_55      4998\n",
       "AAGGAGCAGCGAGAAA-1_56      3355\n",
       "AAGGAGCCACCTCGGA-1_57      2034\n",
       "AAGGAGCCATCGGAAG-1_58      3138\n",
       "AAGGAGCTCGCTTAGA-1_60      1963\n",
       "AAGGCAGCAGGCTCAC-1_61      2686\n",
       "AAGGCAGGTAAGTGGC-1_62      2728\n",
       "AAGGTTCAGCGTCTAT-1_64      1150\n",
       "AAGGTTCGTTTGACAC-1_65      1591\n",
       "AAGGTTCTCAACGAAA-1_66      3616\n",
       "AAGGTTCTCAGCATGT-1_67      7724\n",
       "AAGTCTGAGCAAATCA-1_69      1670\n",
       "AAGTCTGCACCATCCT-1_70      5644\n",
       "AATCCAGCACGAAATA-1_72      2192\n",
       "AATCCAGTCACTCCTG-1_73      4648\n",
       "AATCCAGTCCTTAATC-1_75     11090\n",
       "AATCGGTAGCCAACAG-1_76      1178\n",
       "AATCGGTCACAAGTAA-1_77      7445\n",
       "AATCGGTGTAGAAGGA-1_78      3823\n",
       "AATCGGTTCACTCCTG-1_79      4544\n",
       "ACACCAAAGGTGACCA-1_80      3072\n",
       "ACACCCTAGGCAATTA-1_81      4591\n",
       "ACACCCTCAGTCACTA-1_82      1636\n",
       "ACACCCTGTCTGATTG-1_83      6189\n",
       "ACACCGGAGATCCCGC-1_84      2324\n",
       "ACACCGGTCCTCTAGC-1_85      7108\n",
       "ACAGCCGAGCCATCGC-1_86      1528\n",
       "ACAGCCGAGCCTTGAT-1_87      2609\n",
       "ACAGCCGCAACACGCC-1_88     10111\n",
       "ACAGCCGCAGCGAACA-1_89      2673\n",
       "ACAGCCGTCGGACAAG-1_90      2838\n",
       "ACAGCTAAGATATGCA-1_91      3545\n",
       "ACAGCTAAGCACGCCT-1_92      2503\n",
       "ACAGCTAAGCTAAGAT-1_93      2563\n",
       "ACAGCTAAGGATGCGT-1_94      1185\n",
       "ACAGCTAGTAAGTGGC-1_95      1461\n",
       "ACAGCTAGTGGTGTAG-1_96      3618\n",
       "ACAGCTATCAATCTCT-1_97      1282\n",
       "ACATACGCACCGATAT-1_98      1187\n",
       "ACATACGTCGCTGATA-1_99      1166\n",
       "ACATCAGCAAGGTTTC-1_100     5484\n",
       "ACATCAGCATTTCACT-1_101     1211\n",
       "ACATGGTAGCGTTTAC-1_103     2142\n",
       "ACATGGTAGGAATGGA-1_104     2028\n",
       "ACATGGTGTAAGGATT-1_105     1382\n",
       "ACCAGTACATGCGCAC-1_106     1751\n",
       "ACCAGTACATGGTCAT-1_107     5778\n",
       "ACCAGTAGTTTGTGTG-1_108     1757\n",
       "ACCCACTGTCTGGTCG-1_109     1219\n",
       "ACCGTAAAGAACTGTA-1_110     3407\n",
       "ACCGTAAAGTCCGTAT-1_111     2701\n",
       "ACCTTTAAGAGTGACC-1_112     8181\n",
       "ACCTTTAAGCTTTGGT-1_113     3775\n",
       "Name: n_counts, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The comma separates rows and columns, and the colon returns all rows.\n",
    "cell_attributes.loc[:,'n_counts']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "iloc allows us to subset rows and colums by index number. This is useful if we want to subset multiple rows or columns without typing index names. Lets say we want to remove the columns with names 'orig_ident', 'res_2', and 'louvain'.\n",
    "\n",
    "Lets take a look at the column names first and see if we can slice out the ones we'd like to keep.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['n_counts', 'n_genes', 'percent_mito', 'tree_ident',\n",
       "       'dpt_pseudotime'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Return column names\n",
    "cell_attributes.columns.values\n",
    "cell_attributes.columns.values[[0,1,3,5,7]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "Now we can apply the same indexing pattern to our **iloc** method to return only the columns we're interested in. I've also included a few more slicing variations so you can get a feel for more complex slicing patterns.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>orig_ident</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>res_2</th>\n",
       "      <th>tree_ident</th>\n",
       "      <th>louvain</th>\n",
       "      <th>dpt_pseudotime</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGCACCGCT-1_1</td>\n",
       "      <td>1980</td>\n",
       "      <td>928</td>\n",
       "      <td>1</td>\n",
       "      <td>0.026781</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.353522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCACATTCGA-1_2</td>\n",
       "      <td>2238</td>\n",
       "      <td>894</td>\n",
       "      <td>1</td>\n",
       "      <td>0.024150</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0.177180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCAGGGTTAG-1_3</td>\n",
       "      <td>2839</td>\n",
       "      <td>1025</td>\n",
       "      <td>1</td>\n",
       "      <td>0.005640</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCATGCCTTC-1_4</td>\n",
       "      <td>3437</td>\n",
       "      <td>1438</td>\n",
       "      <td>1</td>\n",
       "      <td>0.012529</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0.103049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTTGAACTC-1_5</td>\n",
       "      <td>2765</td>\n",
       "      <td>1277</td>\n",
       "      <td>1</td>\n",
       "      <td>0.010138</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0.259682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      n_counts  n_genes  orig_ident  percent_mito  res_2  \\\n",
       "index                                                                      \n",
       "AAACCTGAGCACCGCT-1_1      1980      928           1      0.026781      1   \n",
       "AAACCTGCACATTCGA-1_2      2238      894           1      0.024150      6   \n",
       "AAACCTGCAGGGTTAG-1_3      2839     1025           1      0.005640      7   \n",
       "AAACCTGCATGCCTTC-1_4      3437     1438           1      0.012529     10   \n",
       "AAACCTGGTTGAACTC-1_5      2765     1277           1      0.010138      9   \n",
       "\n",
       "                      tree_ident  louvain  dpt_pseudotime  \n",
       "index                                                      \n",
       "AAACCTGAGCACCGCT-1_1           7        0        0.353522  \n",
       "AAACCTGCACATTCGA-1_2           9        2        0.177180  \n",
       "AAACCTGCAGGGTTAG-1_3          14        8        0.000000  \n",
       "AAACCTGCATGCCTTC-1_4          12        1        0.103049  \n",
       "AAACCTGGTTGAACTC-1_5           6       10        0.259682  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Return columns 0, 1, 3, 5, and 7\n",
    "cell_attributes.iloc[:,[0,1,3,5,7]].head(10)\n",
    "\n",
    "# Return rows 1 through 5 and columns 0, 1, 3, 5, and 7\n",
    "cell_attributes.iloc[:5,[0,1,3,5,7]].head(10)\n",
    "\n",
    "# Return rows 1 through 5, columns 1 through 3, and column 7\n",
    "cell_attributes.iloc[:5, 0:3 + 7].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "### Ordering dataframes by column values\n",
    "\n",
    "Here we'll take look at ordering our data by a particular column value, or multiple column values.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>tree_ident</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAGTAGTCACCACCT-1_23</td>\n",
       "      <td>6873</td>\n",
       "      <td>1876</td>\n",
       "      <td>0.018055</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGGTGCAGACA-1_22</td>\n",
       "      <td>3671</td>\n",
       "      <td>1326</td>\n",
       "      <td>0.045244</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAATGCCCACTGCCAG-1_24</td>\n",
       "      <td>1478</td>\n",
       "      <td>680</td>\n",
       "      <td>0.078751</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGAGGTCATCT-1_20</td>\n",
       "      <td>4106</td>\n",
       "      <td>1527</td>\n",
       "      <td>0.017317</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACGGGTCCACGTGG-1_8</td>\n",
       "      <td>5586</td>\n",
       "      <td>2022</td>\n",
       "      <td>0.006445</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGATGCACGAAATA-1_10</td>\n",
       "      <td>1879</td>\n",
       "      <td>822</td>\n",
       "      <td>0.019702</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAATGCCGTATTCGTG-1_25</td>\n",
       "      <td>3129</td>\n",
       "      <td>1107</td>\n",
       "      <td>0.017589</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGCAACACAGATTC-1_15</td>\n",
       "      <td>3206</td>\n",
       "      <td>1263</td>\n",
       "      <td>0.009054</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGGTTGAACTC-1_5</td>\n",
       "      <td>2765</td>\n",
       "      <td>1277</td>\n",
       "      <td>0.010138</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGTCACCTTAT-1_6</td>\n",
       "      <td>2771</td>\n",
       "      <td>1144</td>\n",
       "      <td>0.019530</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGAGCACCGCT-1_1</td>\n",
       "      <td>1980</td>\n",
       "      <td>928</td>\n",
       "      <td>0.026781</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGATGCATCGATTG-1_11</td>\n",
       "      <td>1135</td>\n",
       "      <td>566</td>\n",
       "      <td>0.056537</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGCAAAGATGCCAG-1_14</td>\n",
       "      <td>3953</td>\n",
       "      <td>1414</td>\n",
       "      <td>0.013411</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGATGGTCTTTCAT-1_13</td>\n",
       "      <td>2540</td>\n",
       "      <td>982</td>\n",
       "      <td>0.027964</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGATGGTACACCGC-1_12</td>\n",
       "      <td>1181</td>\n",
       "      <td>542</td>\n",
       "      <td>0.034716</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCACATTCGA-1_2</td>\n",
       "      <td>2238</td>\n",
       "      <td>894</td>\n",
       "      <td>0.024150</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGCAATCGTAGATC-1_18</td>\n",
       "      <td>1647</td>\n",
       "      <td>830</td>\n",
       "      <td>0.019465</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGGTAAGTGGC-1_21</td>\n",
       "      <td>1217</td>\n",
       "      <td>545</td>\n",
       "      <td>0.047658</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCATGCCTTC-1_4</td>\n",
       "      <td>3437</td>\n",
       "      <td>1438</td>\n",
       "      <td>0.012529</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGCAACACGGCGTT-1_16</td>\n",
       "      <td>1355</td>\n",
       "      <td>563</td>\n",
       "      <td>0.056911</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AACACGTGTATCAGTC-1_27</td>\n",
       "      <td>2586</td>\n",
       "      <td>1150</td>\n",
       "      <td>0.006961</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGATGCAAGGACTG-1_9</td>\n",
       "      <td>1904</td>\n",
       "      <td>831</td>\n",
       "      <td>0.025223</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGAGAATAGGG-1_19</td>\n",
       "      <td>1187</td>\n",
       "      <td>544</td>\n",
       "      <td>0.021922</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAACCTGCAGGGTTAG-1_3</td>\n",
       "      <td>2839</td>\n",
       "      <td>1025</td>\n",
       "      <td>0.005640</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AACACGTCAAGCGATG-1_26</td>\n",
       "      <td>1460</td>\n",
       "      <td>671</td>\n",
       "      <td>0.030864</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       n_counts  n_genes  percent_mito  tree_ident\n",
       "index                                                             \n",
       "AAAGTAGTCACCACCT-1_23      6873     1876      0.018055           1\n",
       "AAAGTAGGTGCAGACA-1_22      3671     1326      0.045244           1\n",
       "AAATGCCCACTGCCAG-1_24      1478      680      0.078751           1\n",
       "AAAGTAGAGGTCATCT-1_20      4106     1527      0.017317           2\n",
       "AAACGGGTCCACGTGG-1_8       5586     2022      0.006445           3\n",
       "AAAGATGCACGAAATA-1_10      1879      822      0.019702           3\n",
       "AAATGCCGTATTCGTG-1_25      3129     1107      0.017589           4\n",
       "AAAGCAACACAGATTC-1_15      3206     1263      0.009054           5\n",
       "AAACCTGGTTGAACTC-1_5       2765     1277      0.010138           6\n",
       "AAACCTGTCACCTTAT-1_6       2771     1144      0.019530           7\n",
       "AAACCTGAGCACCGCT-1_1       1980      928      0.026781           7\n",
       "AAAGATGCATCGATTG-1_11      1135      566      0.056537           7\n",
       "AAAGCAAAGATGCCAG-1_14      3953     1414      0.013411           8\n",
       "AAAGATGGTCTTTCAT-1_13      2540      982      0.027964           8\n",
       "AAAGATGGTACACCGC-1_12      1181      542      0.034716           8\n",
       "AAACCTGCACATTCGA-1_2       2238      894      0.024150           9\n",
       "AAAGCAATCGTAGATC-1_18      1647      830      0.019465          11\n",
       "AAAGTAGGTAAGTGGC-1_21      1217      545      0.047658          11\n",
       "AAACCTGCATGCCTTC-1_4       3437     1438      0.012529          12\n",
       "AAAGCAACACGGCGTT-1_16      1355      563      0.056911          12\n",
       "AACACGTGTATCAGTC-1_27      2586     1150      0.006961          13\n",
       "AAAGATGCAAGGACTG-1_9       1904      831      0.025223          13\n",
       "AAAGTAGAGAATAGGG-1_19      1187      544      0.021922          13\n",
       "AAACCTGCAGGGTTAG-1_3       2839     1025      0.005640          14\n",
       "AACACGTCAAGCGATG-1_26      1460      671      0.030864          14"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's make a smaller dataset to work with\n",
    "cell_df_sub = cell_attributes.iloc[:25,[0,1,3,5]]\n",
    "\n",
    "# Set ascending=True to reverse the order\n",
    "cell_df_sub.sort_values('n_counts', ascending=False)\n",
    "\n",
    "# Sort by multiple columns in different directions\n",
    "cell_df_sub.sort_values(by=['tree_ident', 'n_counts'], ascending=[True, False])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "### Subsetting data by condition\n",
    "\n",
    "Understanding how to subset your data using conditional operations is very, _very_ useful. You'll often encounter situations where you want to filter your data on a certain set of parameters to reduce it to a more \"meaningful\" state (to make your PI happy).\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>tree_ident</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAGTAGGTGCAGACA-1_22</td>\n",
       "      <td>3671</td>\n",
       "      <td>1326</td>\n",
       "      <td>0.045244</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGTCACCACCT-1_23</td>\n",
       "      <td>6873</td>\n",
       "      <td>1876</td>\n",
       "      <td>0.018055</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAATGCCCACTGCCAG-1_24</td>\n",
       "      <td>1478</td>\n",
       "      <td>680</td>\n",
       "      <td>0.078751</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       n_counts  n_genes  percent_mito  tree_ident\n",
       "index                                                             \n",
       "AAAGTAGGTGCAGACA-1_22      3671     1326      0.045244           1\n",
       "AAAGTAGTCACCACCT-1_23      6873     1876      0.018055           1\n",
       "AAATGCCCACTGCCAG-1_24      1478      680      0.078751           1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subsetting on a single condition\n",
    "cell_df_sub.loc[(cell_df_sub['tree_ident'] == 1),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "In the second example we're chaining together boolean operators to achieve results that satisfy multiple conditions. You can make these statments complex as you'd like.\n",
    "\n",
    "Note: Pandas uses a pipe symbol to represent \"or\", and an ampersand symbol to represent \"and\". The backslashes in code simply allow us to break up our statement at arbitrary points for readbility.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>tree_ident</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAGTAGAGGTCATCT-1_20</td>\n",
       "      <td>4106</td>\n",
       "      <td>1527</td>\n",
       "      <td>0.017317</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGGTGCAGACA-1_22</td>\n",
       "      <td>3671</td>\n",
       "      <td>1326</td>\n",
       "      <td>0.045244</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAGTAGTCACCACCT-1_23</td>\n",
       "      <td>6873</td>\n",
       "      <td>1876</td>\n",
       "      <td>0.018055</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAATGCCCACTGCCAG-1_24</td>\n",
       "      <td>1478</td>\n",
       "      <td>680</td>\n",
       "      <td>0.078751</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       n_counts  n_genes  percent_mito  tree_ident\n",
       "index                                                             \n",
       "AAAGTAGAGGTCATCT-1_20      4106     1527      0.017317           2\n",
       "AAAGTAGGTGCAGACA-1_22      3671     1326      0.045244           1\n",
       "AAAGTAGTCACCACCT-1_23      6873     1876      0.018055           1\n",
       "AAATGCCCACTGCCAG-1_24      1478      680      0.078751           1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Subsetting on multiple conditions.\n",
    "cell_df_sub.loc[\n",
    "    (cell_df_sub['tree_ident'] == 1) | \\\n",
    "    (cell_df_sub['tree_ident'] == 2) & \\\n",
    "    (cell_df_sub['n_genes'] > 1000),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br/>\n",
    "\n",
    "What's actually going on here? The rows in the data frame are actually subsetted on a vector of True/False statements. That is, for every condition that is True for all statements, a row will be returned. If remove the boonlean statements placed within cell_df_sub.loc[], you can see why this is occuring.\n",
    "\n",
    "<br/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "AAACCTGAGCACCGCT-1_1     False\n",
       "AAACCTGCACATTCGA-1_2     False\n",
       "AAACCTGCAGGGTTAG-1_3     False\n",
       "AAACCTGCATGCCTTC-1_4     False\n",
       "AAACCTGGTTGAACTC-1_5     False\n",
       "AAACCTGTCACCTTAT-1_6     False\n",
       "AAACGGGTCCACGTGG-1_8     False\n",
       "AAAGATGCAAGGACTG-1_9     False\n",
       "AAAGATGCACGAAATA-1_10    False\n",
       "AAAGATGCATCGATTG-1_11    False\n",
       "AAAGATGGTACACCGC-1_12    False\n",
       "AAAGATGGTCTTTCAT-1_13    False\n",
       "AAAGCAAAGATGCCAG-1_14    False\n",
       "AAAGCAACACAGATTC-1_15    False\n",
       "AAAGCAACACGGCGTT-1_16    False\n",
       "AAAGCAATCGTAGATC-1_18    False\n",
       "AAAGTAGAGAATAGGG-1_19    False\n",
       "AAAGTAGAGGTCATCT-1_20    False\n",
       "AAAGTAGGTAAGTGGC-1_21    False\n",
       "AAAGTAGGTGCAGACA-1_22     True\n",
       "AAAGTAGTCACCACCT-1_23     True\n",
       "AAATGCCCACTGCCAG-1_24     True\n",
       "AAATGCCGTATTCGTG-1_25    False\n",
       "AACACGTCAAGCGATG-1_26    False\n",
       "AACACGTGTATCAGTC-1_27    False\n",
       "dtype: bool"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_df_sub['tree_ident'] == 1 | \\\n",
    "    (cell_df_sub['tree_ident'] == 2) & \\\n",
    "    (cell_df_sub['n_genes'] > 1000)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
